src/main_chat/chatbot_manager.py

```diff
@@ -1,45 +1,46 @@
-
-    def _is_general_chatbot_unemployed(self):
-        return self._bot_context.is_after_greeting and \
-               self._bot_context.is_after_introduction and \
-               self._bot_context.is_name_request_processed and \
-               self._bot_context.is_after_name_response_reaction and \
-               self._bot_context.speaker_name
-
-    def _ask_general_chatbot(self, processed_sentence):
-        response = self._general_chatbot.get_response(processed_sentence)
-        self._bot_context.context_update(response.in_response_to)
-        return response.text
+    def _ask_intro_chatbot(self, processed_sentence):
+        response = self._intro_chatbot.get_bot_response(processed_sentence)
+        self._intro_chatbot.context_update(response.in_response_to)
+        return response.text, response.confidence
 
     def _ask_university_chatbot(self, processed_sentence):
-        response = self._university_chatbot.get_response(processed_sentence)
-        return response.text
-
-    def ask_chatbot(self, user_input): # this is key method which is called from main.py
-        if self._is_general_chatbot_unemployed():
-            processed_sentence = self._sentence_filter.filter_sentence(user_input)
-            print(processed_sentence)
-            chatbot_response = self._ask_university_chatbot(processed_sentence)
+        response = self._university_chatbot.get_bot_response(processed_sentence)
+        return response.text, response.confidence
+
+    def _check_is_intro_chatbot_unemployed(self):
+        if not self._is_intro_bot_unemployed:
+            self._is_intro_bot_unemployed = self._intro_chatbot.check_is_bot_unemployed() \
+                                            or self._university_chatbot.check_was_requested_in_row_above_thresh()
+        return self._is_intro_bot_unemployed
+
+    def check_is_intro_chatbot_part_employed(self):
+        return self._intro_chatbot.check_is_bot_partially_employed()
+
+    def ask_chatbot(self, user_input):  # this is key method which is called from main.py
+        if self._check_is_intro_chatbot_unemployed():
+            chatbot_response, c1 = self._ask_university_chatbot(user_input)
+            print('University chatbot = ', user_input, ' c1 = ', c1)
         else:
-            print(user_input)
-            chatbot_response = self._ask_general_chatbot(user_input)
-
-        return chatbot_response
\ No newline at end of file
+            (i_text, i_conf) = self._ask_intro_chatbot(user_input)
+            (u_text, u_conf) = self._ask_university_chatbot(user_input)
+            print("U_Text = {}, u_conf = {}".format(u_text, u_conf))
+            conf_res = u_conf > i_conf
+            self._university_chatbot.inc_responses_in_row() if conf_res else self._university_chatbot.reset_responses_in_row()
+            chatbot_response = u_text if conf_res else i_text
+        return chatbot_response
```

Comment:
I think it would be better to implement this module as more general, now it is too much tied to assumption that we use only two bots, and adding third one when needed will cause code changes.

Reply:
yes, but for now i assumed that we have these 2 bots and refactoring to enable adding new bots will be done in next steps of our work. But i agree with you to make this class more general.